numba/core/event.py

```diff
@@ -0,0 +1,357 @@
+
+    @property
+    def done(self):
+        """Returns a ``bool`` indicating a measurement is made.
+
+        When this returns ``False``, the matching event has never fired.
+        Iff this returns ``True``, ``.duration`` can be read without error.
+        """
+        return hasattr(self, "_duration")
+
+    @property
+    def duration(self):
+        """Returns the measured duration.
+
+        This may raise ``AttributeError``. Users can use ``.done`` to check
+        that a measurement has been made.
+        """
+        return self._duration
+
+
+class RecordingListener(Listener):
+    """A listener that records all event and store it in the ``.buffer``
+    attribute as a list of 2-tuple ``(float, Event)``, where the first element
+    is the time of the event as returned by ``time.time()`` and the second
+    element is the event.
+    """
+    def __init__(self):
+        self.buffer = []
+
+    def on_start(self, event):
+        self.buffer.append((time.time(), event))
+
+    def on_end(self, event):
+        self.buffer.append((time.time(), event))
+
+
+@contextmanager
+def install_listener(kind, listener):
+    """Install a listener temporarily within the duration of the context.
+
+    Returns
+    -------
+    res : Listener
+        The *listener* provided.
+    """
+    register(kind, listener)
+    try:
+        yield listener
+    finally:
+        unregister(kind, listener)
+
+
+@contextmanager
+def install_timer(kind, callback):
+    """Install a TimingListener temporarily to measure the duration for
+    an event.
+
+    If the context completes successfully, the *callback* function is executed.
+    The *callback* function is expected to take a float argument for the
+    duration in seconds.
+
+    Returns
+    -------
+    res : TimingListener
+    """
+    tl = TimingListener()
+    with install_listener(kind, tl):
+        yield tl
+
+    if tl.done:
+        callback(tl.duration)
+
+
+@contextmanager
+def install_recorder(kind):
+    """Install a RecordingListener temporarily to record all events.
+
+    The buffer is filled from the context is closed.
+
+    Returns
+    -------
+    res : RecordingListener
+    """
+    rl = RecordingListener()
+    with install_listener(kind, rl):
+        yield rl
+
+
+def start_event(kind, data=None):
+    """Signal the start of an event of *kind* with *data*.
+
+    Parameters
+    ----------
+    kind : str
+        Event kind.
+    data : any; optional
+        Extra event data.
+    """
+    evt = Event(kind=kind, status=EventStatus.START, data=data)
+    broadcast(evt)
+
+
+def end_event(kind, data=None, exc_details=None):
+    """Signal the end of an event of *kind*, *exc_details*.
+
+    Parameters
+    ----------
+    kind : str
+        Event kind.
+    data : any; optional
+        Extra event data.
+    """
+    evt = Event(
+        kind=kind, status=EventStatus.END, data=data, exc_details=exc_details,
+    )
+    broadcast(evt)
+
```

Comment:
In addition to the plain `start_event` and `end_event` it may be useful to wrap the builtin kinds here for API convenience. like `start_compiler_lock` and `end_compiler_lock` as well as `start_compile` and `end_compile`. What do you think.

Reply:
I don't think the extra wrappers are saving that much. I also don't want too many ways to do the same thing. Most cases can use the `mark_event()` context manager, which is only one call instead of two. I'm deliberately keep it as simple and minimal as possible.

Reply:
OK, makes sense.